File: order-service/src/main/java/github/lms/lemuel/product/domain/ProductVariant.java
package github.lms.lemuel.product.domain;
import github.lms.lemuel.product.domain.exception.InvalidProductStateException;
import github.lms.lemuel.product.domain.exception.ProductInvariantViolationException;

import github.lms.lemuel.product.domain.exception.InsufficientStockException;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.util.Objects;

/**
 * 상품 옵션 (SKU / Variant) 도메인.
 *
 * <p>색상·사이즈 등 옵션 조합 1 개 = SKU 1 개. 옵션 상품의 재고는 {@link Product} 가 아닌
 * 이 객체에서 관리되며, 결제·주문은 productId 가 아닌 variantId(또는 sku) 를 기준으로 동작한다.
 *
 * <p>동시성 정책:
 * <ul>
 *   <li>재고 차감은 {@code DecreaseVariantStockService} 의 원자적 조건부 UPDATE 로 처리 →
 *       동시 차감 폭주에도 락 대기·재시도 없이 초과판매 방지</li>
 *   <li>{@code version} 필드는 {@code @Version} 으로 매핑되어 일반 부분 수정(옵션/가격 변경 등)의
 *       lost update 를 막는다. 차감 경로는 UPDATE 문에서 version 을 직접 +1 한다.</li>
 * </ul>
 */
public class ProductVariant {

    private Long id;
    private final Long productId;
    private final String sku;
    private String optionName;
    private BigDecimal additionalPrice;
    private BigDecimal discountPrice;
    private BigDecimal discountRate;
    /** 이 SKU 를 사 오는 값. null 은 "아직 모른다"이고 0원 매입이 아니다. */
    private BigDecimal purchasePrice;
    private int stockQuantity;
    private long version;
    private ProductVariantStatus status;
    private String optionSignature;
    private final LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    public static ProductVariant create(Long productId, String sku, String optionName,
                                         BigDecimal additionalPrice, int initialStock) {
        Objects.requireNonNull(productId, "productId");
        Objects.requireNonNull(sku, "sku");
        if (sku.isBlank()) throw new ProductInvariantViolationException("sku 는 필수");
        if (optionName == null || optionName.isBlank()) {
            throw new ProductInvariantViolationException("optionName 은 필수 (예: '색상:빨강/사이즈:L')");
        }
        if (initialStock < 0) {
            throw new ProductInvariantViolationException("초기 재고는 0 이상");
        }
        BigDecimal price = additionalPrice == null ? BigDecimal.ZERO : additionalPrice;
        return new ProductVariant(null, productId, sku, optionName, price, null, null, initialStock,
                0L, initialStock == 0 ? ProductVariantStatus.OUT_OF_STOCK : ProductVariantStatus.ACTIVE,
                LocalDateTime.now(), LocalDateTime.now());
    }

    public static ProductVariant rehydrate(Long id, Long productId, String sku, String optionName,
                                            BigDecimal additionalPrice, int stockQuantity, long version,
                                            ProductVariantStatus status, LocalDateTime createdAt,
                                            LocalDateTime updatedAt) {
        return rehydrate(id, productId, sku, optionName, additionalPrice, null, null, stockQuantity,
                version, status, createdAt, updatedAt);
    }

    public static ProductVariant rehydrate(Long id, Long productId, String sku, String optionName,
                                            BigDecimal additionalPrice, BigDecimal discountPrice,
                                            BigDecimal discountRate, int stockQuantity, long version,
                                            ProductVariantStatus status, LocalDateTime createdAt,
                                            LocalDateTime updatedAt) {
        return rehydrate(id, productId, sku, optionName, additionalPrice, discountPrice, discountRate,
                stockQuantity, version, status, null, createdAt, updatedAt);
    }

    public static ProductVariant rehydrate(Long id, Long productId, String sku, String optionName,
                                            BigDecimal additionalPrice, BigDecimal discountPrice,
                                            BigDecimal discountRate, int stockQuantity, long version,
                                            ProductVariantStatus status, String optionSignature,
                                            LocalDateTime createdAt, LocalDateTime updatedAt) {
        return rehydrate(id, productId, sku, optionName, additionalPrice, discountPrice, discountRate,
                null, stockQuantity, version, status, optionSignature, createdAt, updatedAt);
    }

    public static ProductVariant rehydrate(Long id, Long productId, String sku, String optionName,
                                            BigDecimal additionalPrice, BigDecimal discountPrice,
                                            BigDecimal discountRate, BigDecimal purchasePrice,
                                            int stockQuantity, long version,
                                            ProductVariantStatus status, String optionSignature,
                                            LocalDateTime createdAt, LocalDateTime updatedAt) {
        ProductVariant variant = new ProductVariant(id, productId, sku, optionName, additionalPrice,
                discountPrice, discountRate, stockQuantity, version, status, createdAt, updatedAt);
        variant.optionSignature = optionSignature;
        variant.purchasePrice = purchasePrice;
        return variant;
    }

    private ProductVariant(Long id, Long productId, String sku, String optionName,
                           BigDecimal additionalPrice, BigDecimal discountPrice, BigDecimal discountRate,
                           int stockQuantity, long version, ProductVariantStatus status,
                           LocalDateTime createdAt, LocalDateTime updatedAt) {
        this.id = id;
        this.productId = productId;
        this.sku = sku;
        this.optionName = optionName;
        this.additionalPrice = additionalPrice;
        this.discountPrice = discountPrice;
        this.discountRate = discountRate;
        this.stockQuantity = stockQuantity;
        this.version = version;
        this.status = status;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    /**
     * 재고 차감 도메인 불변식 가드(음수 재고 방지). 고동시성 차감은 영속 계층의 원자적 조건부
     * UPDATE 가 담당하므로, 이 메서드는 단건/검증 용도로만 사용한다.
     */
    public void decreaseStock(int quantity) {
        if (quantity <= 0) {
            throw new ProductInvariantViolationException("차감 수량은 양수여야 합니다");
        }
        if (this.status == ProductVariantStatus.DISCONTINUED) {
            throw new InvalidProductStateException("단종된 SKU 는 차감할 수 없습니다: " + sku);
        }
        if (this.stockQuantity < quantity) {
            throw new InsufficientStockException(
                    "재고 부족: sku=" + sku + ", 요청=" + quantity + ", 가용=" + stockQuantity);
        }
        this.stockQuantity -= quantity;
        if (this.stockQuantity == 0) {
            this.status = ProductVariantStatus.OUT_OF_STOCK;
        }
        this.updatedAt = LocalDateTime.now();
    }

    /**
     * 이 옵션이 적용된 <b>주문 단가</b>를 계산한다 (할인 적용 후).
     *
     * <p>금액 우선순위(피드백 합의 — 환불 금액 역산도 이 순서를 따른다):
     * <ol>
     *   <li>기준 가격({@code basePrice}) — products.price 스냅샷</li>
     *   <li>+ 옵션 추가금({@code additionalPrice}, 음수 가능)</li>
     *   <li>- 옵션 정액 할인({@code discountPrice})</li>
     *   <li>- 옵션 정률 할인({@code discountRate} %) — 위 (기준가+추가금-정액할인) 에 적용, 원 단위 버림(FLOOR)</li>
     * </ol>
     * 최종 단가가 음수가 되면 0 으로 절삭한다. 두 할인 필드는 null 이면 미적용.
     */
    public BigDecimal effectiveUnitPrice(BigDecimal basePrice) {
        Objects.requireNonNull(basePrice, "basePrice");
        BigDecimal price = basePrice.add(additionalPrice == null ? BigDecimal.ZERO : additionalPrice);
        if (discountPrice != null) {
            price = price.subtract(discountPrice);
        }
        if (discountRate != null && price.signum() > 0) {
            BigDecimal rateDiscount = price.multiply(discountRate)
                    .divide(BigDecimal.valueOf(100), 0, RoundingMode.FLOOR);
            price = price.subtract(rateDiscount);
        }
        return price.signum() < 0 ? BigDecimal.ZERO : price;
    }

    /**
     * 매입가를 정하거나 고친다. {@code null} 을 넣으면 "모른다"로 되돌린다 — 잘못 넣은 값을
     * 0 으로 덮는 것과 지우는 것은 다른 일이라, 지울 길을 막아 두면 0원 매입이라는 거짓이 남는다.
     *
     * @throws ProductInvariantViolationException 음수를 넣은 경우. 역마진은 허용하지만
     *         (그건 판매가와의 관계이지 이 값의 문제가 아니다) 음수 매입가는 입력 사고다.
     */
    public void changePurchasePrice(BigDecimal newPurchasePrice) {
        if (newPurchasePrice != null && newPurchasePrice.signum() < 0) {
            throw new ProductInvariantViolationException("매입가는 0 이상이어야 합니다: " + newPurchasePrice);
        }
        this.purchasePrice = newPurchasePrice;
        this.updatedAt = LocalDateTime.now();
    }

    public boolean hasPurchasePrice() {
        return purchasePrice != null;
    }

    /**
     * 이 SKU 1 개를 팔았을 때 남는 금액 = 판매가 - 매입가.
     *
     * <p>판매가는 {@link #effectiveUnitPrice(BigDecimal)} 로 <b>그때그때 계산한</b> 값을 쓴다.
     * 기준가·추가금·할인 중 무엇이 바뀌어도 이 값이 따라 움직여야 하기 때문이다. 마진을 컬럼으로
     * 굳혀 두지 않은 이유가 여기 있다 — 굳힌 값은 넷 중 하나만 바뀌어도 조용히 거짓이 된다.
     *
     * <p>매입가를 모르면 {@code null} 이다. 0 이 아니다. 모르는 것을 0 으로 답하면 마진이
     * 판매가 전액으로 잡혀 리포트가 통째로 부풀어 오른다.
     *
     * <p>결과는 음수일 수 있다(역마진). 깎지 않는다 — 손해 보고 파는 SKU 는 가려야 할 것이 아니라
     * 눈에 띄어야 하는 것이다.
     */
    public BigDecimal marginAmount(BigDecimal basePrice) {
        if (purchasePrice == null) {
            return null;
        }
        return effectiveUnitPrice(basePrice).subtract(purchasePrice);
    }

    /**
     * 마진율(%) = 마진액 / 판매가 × 100. 소수점 둘째 자리에서 반올림한다.
     *
     * <p>분모는 매입가가 아니라 <b>판매가</b>다. 즉 "판 값의 몇 %가 남았나"(매출 총이익률)이지
     * "산 값 대비 몇 % 붙였나"(마크업)가 아니다. 두 숫자는 같은 거래에서도 다르게 나온다 —
     * 1000원에 사서 2000원에 팔면 이익률 50%, 마크업 100%. 어느 쪽인지 적어 두지 않으면
     * 보는 사람마다 다르게 읽는다.
     *
     * <p>매입가를 모르거나 판매가가 0 이면 {@code null} 이다. 0 으로 나눌 수 없고, 0원에 판
     * 물건의 이익률은 정의되지 않는다(무료 증정이 마진 -∞ 로 리포트에 찍히면 곤란하다).
     */
    public BigDecimal marginRate(BigDecimal basePrice) {
        BigDecimal margin = marginAmount(basePrice);
        if (margin == null) {
            return null;
        }
        BigDecimal sellingPrice = effectiveUnitPrice(basePrice);
        if (sellingPrice.signum() == 0) {
            return null;
        }
        return margin.multiply(BigDecimal.valueOf(100))
                .divide(sellingPrice, 2, RoundingMode.HALF_UP);
    }

    public void increaseStock(int quantity) {
        if (quantity <= 0) {
            throw new ProductInvariantViolationException("증가 수량은 양수여야 합니다");
        }
        this.stockQuantity += quantity;
        if (this.status == ProductVariantStatus.OUT_OF_STOCK && stockQuantity > 0) {
            this.status = ProductVariantStatus.ACTIVE;
        }
        this.updatedAt = LocalDateTime.now();
    }

    public void discontinue() {
        this.status = ProductVariantStatus.DISCONTINUED;
        this.updatedAt = LocalDateTime.now();
    }

    public boolean isAvailable() {
        return status == ProductVariantStatus.ACTIVE && stockQuantity > 0;
    }

    /**
     * 조합 서명 부여 — 이 SKU 가 어떤 옵션 값 조합인지의 기계 식별자({@link OptionSignature}).
     *
     * <p>같은 값을 다시 넣는 것은 no-op 이라 백필을 몇 번 돌려도 안전하다. 반면 <b>다른</b> 서명을
     * 덮어쓰려는 것은 "이 SKU 의 옵션 조합이 바뀌었다"는 뜻인데, 조합이 바뀌면 그건 다른 SKU 다 —
     * 이미 팔린 주문이 가리키는 조합이 소급해서 달라지므로 막는다.
     */
    public void assignOptionSignature(String signature) {
        Objects.requireNonNull(signature, "signature");
        if (this.optionSignature != null && !this.optionSignature.equals(signature)) {
            throw new InvalidProductStateException(
                    "이미 다른 조합 서명이 부여된 SKU 입니다: sku=" + sku);
        }
        this.optionSignature = signature;
        this.updatedAt = LocalDateTime.now();
    }

    public boolean hasOptionSignature() {
        return optionSignature != null;
    }

    public Long getId() { return id; }
    public Long getProductId() { return productId; }
    public String getSku() { return sku; }
    public String getOptionName() { return optionName; }
    public BigDecimal getAdditionalPrice() { return additionalPrice; }
    public BigDecimal getDiscountPrice() { return discountPrice; }
    public BigDecimal getDiscountRate() { return discountRate; }
    public BigDecimal getPurchasePrice() { return purchasePrice; }
    public int getStockQuantity() { return stockQuantity; }
    public long getVersion() { return version; }
    public ProductVariantStatus getStatus() { return status; }
    public String getOptionSignature() { return optionSignature; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public LocalDateTime getUpdatedAt() { return updatedAt; }

    /**
     * Persistence 어댑터에서 INSERT 후 생성된 PK 를 주입할 때만 사용.
     */
    public void assignId(Long id) {
        if (this.id != null) {
            throw new IllegalStateException("id 는 1 회만 부여 가능");
        }
        this.id = id;
    }
}

File: frontend/src/api/productVariant.ts
import api from './axios';

/**
 * 상품 옵션(SKU) — order-service {@code ProductVariantController}.
 *
 * <p><b>왜 생겼나.</b> 옵션 축과 값을 정의하는 콘솔은 있었다({@code /admin/system/option-catalog}).
 * 하지만 그건 "색상에는 빨강·파랑이 있다"는 <b>사전</b>이고, "이 상품의 빨강/L 은 재고 3개,
 * 추가금 2,000원"이라는 <b>실물</b>을 만드는 화면은 없었다. 그래서 SKU 를 만드는 길은
 * {@code curl} 뿐이었고, 재고가 어긋나면 DB 를 직접 고쳤다.
 *
 * <h3>응답이 {@code {"variant": {...}}} 로 한 겹 싸여 있다</h3>
 * <p>서버가 필드 순서를 보존하면서 <b>null 을 허용</b>하려고 맵으로 담아 내보내는 구조라,
 * 응답이 평평한 객체가 아니다. 화면이 이 껍질을 안 벗기면 모든 필드가 {@code undefined} 로
 * 조용히 비어 렌더된다 — 에러가 아니라 <b>빈 표</b>가 되므로 눈에 잘 안 띈다. 벗기는 자리를
 * 여기 한 곳으로 모은다.
 *
 * <h3>재고 차감은 화면이 함부로 부를 것이 아니다</h3>
 * <p>{@link productVariantApi.decreaseStock} 은 주문이 프로세스 안에서 부르는 것과 <b>같은</b>
 * 유스케이스를 HTTP 로 여는 문이다. 즉 이 버튼을 누르면 주문 없이 재고가 준다. 정정·폐기처럼
 * 실물이 사라진 경우를 반영하는 운영자 조작이고, 서버도 ADMIN 으로만 연다
 * ({@code SecurityConfig} 의 {@code POST /products/*&#47;variants/*&#47;decrease-stock}).
 * 되돌리는 API 는 없다 — 늘리려면 SKU 를 다시 만드는 수밖에 없다.
 */

export type VariantStatus = string;

/** 껍질을 벗긴 SKU 한 줄. 할인 두 필드는 미설정이 정상이다(대부분의 SKU 가 할인 없음). */
export interface ProductVariant {
  id: number;
  productId: number;
  sku: string;
  optionName: string;
  additionalPrice: number | null;
  discountPrice: number | null;
  discountRate: number | null;
  stockQuantity: number;
  /** 낙관적 락 버전. 화면은 읽기만 한다 — 충돌 재시도는 서버가 한다. */
  version: number;
  status: VariantStatus;
  /** 옵션 트리 선택 경로를 정규화한 서명. 트리에서 SKU 를 되찾는 열쇠다. */
  optionSignature: string | null;
}

/** 서버가 내보내는 실제 모양. 화면은 이 타입을 밖으로 내보내지 않는다. */
interface VariantEnvelope {
  variant: ProductVariant;
}

export interface CreateVariantRequest {
  sku: string;
  optionName: string;
  /** 기본가에 더할 금액. 비워 두면 서버가 0 으로 본다. */
  additionalPrice?: number | null;
  initialStock: number;
}

/** 옵션 트리에서 고른 한 칸. 이름과 값이 둘 다 서버의 정의와 맞아야 해석된다. */
export interface OptionSelection {
  name: string;
  value: string;
}

const unwrap = (envelope: VariantEnvelope): ProductVariant => envelope.variant;

// 경로는 전체 리터럴로 적는다. 조각을 이어 붙이면 사람 눈에도, 저장소의 화면-API 대조
// 게이트(api-screen-gate)에도 어떤 엔드포인트를 부르는지 보이지 않는다.
export const productVariantApi = {
  /** GET — 그 상품의 SKU 전부. */
  list: async (productId: number): Promise<ProductVariant[]> =>
    (await api.get<VariantEnvelope[]>(`/products/${productId}/variants`)).data.map(unwrap),

  /** POST — SKU 생성. sku 는 상품 안에서 유일해야 한다(중복이면 서버가 거절). */
  create: async (productId: number, body: CreateVariantRequest): Promise<ProductVariant> =>
    unwrap((await api.post<VariantEnvelope>(`/products/${productId}/variants`, body)).data),

  /** POST — 재고 차감. 주문 없이 재고가 줄어드는 조작이라 되돌릴 수 없다. */
  decreaseStock: async (
    productId: number,
    variantId: number,
    quantity: number,
  ): Promise<ProductVariant> =>
    unwrap((await api.post<VariantEnvelope>(
      `/products/${productId}/variants/${variantId}/decrease-stock`,
      { quantity },
    )).data),

  /** POST — 옵션 선택 경로를 SKU 하나로 해석한다. 주문이 variantId 를 얻는 길과 같은 경로다. */
  resolve: async (productId: number, selections: OptionSelection[]): Promise<ProductVariant> =>
    unwrap((await api.post<VariantEnvelope>(
      `/products/${productId}/variants/resolve`,
      { selections },
    )).data),
};

/**
 * SKU 한 줄의 원가와 마진 — order-service {@code AdminVariantCostController}.
 *
 * <p><b>왜 위 {@link ProductVariant} 에 필드로 붙이지 않았나.</b> 그 타입은 구매자가 부르는
 * {@code POST /products/{id}/variants/resolve} 의 응답이기도 하다. 거기에 원가를 얹으면
 * 로그인한 아무나 원가를 본다. 서버가 응답을 갈라 놓았으므로 화면 타입도 갈라 둔다 —
 * 한쪽으로 합치는 순간 그 사실이 보이지 않게 된다.
 *
 * <p>{@code marginAmount}·{@code marginRate} 가 {@code null} 인 것은 마진이 0 이라는 뜻이
 * 아니라 <b>매입가를 아직 모른다</b>는 뜻이다. 화면은 이 둘을 다르게 그려야 한다.
 */
export interface VariantCost {
  variantId: number;
  sku: string;
  optionName: string;
  stockQuantity: number;
  /** 저장된 값이 아니라 기준가+추가금-할인으로 서버가 계산한 값. */
  sellingPrice: number;
  /** null 은 미입력. 0 은 "0원에 샀다"로 다른 뜻이다. */
  purchasePrice: number | null;
  /** 판매가 - 매입가. 역마진이면 음수 그대로 온다. */
  marginAmount: number | null;
  /** 판매가 대비 마진율(매출총이익률, %). 매입가 대비 가산율이 아니다. */
  marginRate: number | null;
}

export const variantCostApi = {
  /** GET — 그 상품 SKU 들의 원가·마진. 관리자 전용 경로다. */
  list: async (productId: number): Promise<VariantCost[]> =>
    (await api.get<VariantCost[]>(`/admin/products/${productId}/variants/costs`)).data,

  /** PUT — 매입가 설정. null 을 보내면 "모른다"로 되돌린다(0 으로 덮는 것과 다르다). */
  setPurchasePrice: async (
    productId: number,
    variantId: number,
    purchasePrice: number | null,
  ): Promise<VariantCost> =>
    (await api.put<VariantCost>(
      `/admin/products/${productId}/variants/${variantId}/purchase-price`,
      { purchasePrice },
    )).data,
};

/**
 * 화면에 보일 판매가 — 기본가 + 추가금, 할인가가 있으면 그것.
 *
 * <p>할인가는 "기본가에서 얼마를 뺀 값"이 아니라 <b>확정된 판매가</b>다. 그래서 있으면
 * 추가금을 더하지 않고 그대로 쓴다. 두 값을 겹쳐 계산하면 화면의 금액이 결제 금액과 갈린다.
 */
export function effectivePrice(basePrice: number, variant: ProductVariant): number {
  if (variant.discountPrice != null) return variant.discountPrice;
  return basePrice + (variant.additionalPrice ?? 0);
}
